### PR TITLE
Update ingress/terminating gateway stats labeling

### PR DESCRIPTION
**Pull request by freddygv**
Fixes: #9887

This PR updates the stats_prefix for ingress and terminating gateways.

1. For ingress gateways the change makes it so that the listener _port_ (`listenerKey.RouteName()`) does not get encoded in the metric name. (Brought up in the issue)
2. For terminating gateways we also make it so that the listener _name_ does not get encoded in the metric name. There was also a change to the filter name so that the rules defined [here](https://github.com/hashicorp/consul/blob/64d122b0a2447a6e0597bdaa3912e83cd15ef3ed/command/connect/envoy/bootstrap_config.go#L392:L410) apply. This will enable per-destination listener metrics.

These changes are **NOT** backwards compatible.

### Example from issue:

In 1.9.x:

```
# envoy stat: http.ingress_upstream.14146.no_route: 0
envoy_http_14146_no_route{
  local_cluster="ingress-service",
  consul_source_service="ingress-service",
  consul_source_namespace="default",
  consul_source_datacenter="dc1",
  envoy_http_conn_manager_prefix="ingress_upstream"
} 0
```
```
envoy_tcp_default_web_default_downstream_cx_no_route{
  local_cluster="terminating-gateway
  consul_source_service="terminating-gateway
  consul_source_namespace="default
  consul_source_datacenter="dc1
  envoy_tcp_prefix="terminating_gateway"
} 0
```

Proposed in PR:

```
envoy_http_no_route{
  local_cluster="ingress-service",
  consul_source_service="ingress-service",
  consul_source_namespace="default",
  consul_source_datacenter="dc1",
  envoy_http_conn_manager_prefix="ingress_upstream_14146"
} 0
```

```
envoy_tcp_downstream_cx_no_route{
  local_cluster="terminating-gateway",
  consul_source_service="terminating-gateway",
  consul_source_namespace="default",
  consul_source_datacenter="dc1",
  consul_upstream_service="web",
  consul_upstream_datacenter="dc1",
  consul_upstream_namespace="default",
  envoy_tcp_prefix="upstream"
} 0
```

### CODE REVIEW

File: agent/xds/testdata/listeners/terminating-gateway-custom-and-tagged-addresses.envoy-1-18-x.golden [hunk 1]
```diff
@@ -32,7 +32,7 @@
               "name": "envoy.filters.network.tcp_proxy",
               "typedConfig": {
                 "@type": "type.googleapis.com/envoy.extensions.filters.network.tcp_proxy.v3.TcpProxy",
-                "statPrefix": "terminating_gateway.default.api.foo",
+                "statPrefix": "upstream.api.default.dc1.",
```
Comment (rboyer) [inline agent/xds/testdata/listeners/terminating-gateway-custom-and-tagged-addresses.envoy-1-18-x.golden:35]: is this supposed to have a trailing dot?

File: agent/xds/listeners.go [hunk 2]
```diff
@@ -1133,13 +1131,12 @@ func (s *ResourceGenerator) makeFilterChainTerminatingGateway(
 	// Lastly we setup the actual proxying component. For L4 this is a straight
 	// tcp proxy. For L7 this is a very hands-off HTTP proxy just to inject an
 	// HTTP filter to do intention checks here instead.
-	statPrefix := fmt.Sprintf("terminating_gateway.%s.%s.", service.NamespaceOrDefault(), service.Name)
 	opts := listenerFilterOpts{
 		protocol:   protocol,
-		filterName: listener,
+		filterName: fmt.Sprintf("%s.%s.%s.", service.Name, service.NamespaceOrDefault(), cfgSnap.Datacenter),
```
Comment (rboyer) [inline agent/xds/listeners.go:1136]:
```suggestion
		filterName: fmt.Sprintf("%s.%s.%s", service.Name, service.NamespaceOrDefault(), cfgSnap.Datacenter),
```
This is the actual filter name, not a prefix, so I don't think it should get a trailing dot.

File: agent/xds/testdata/listeners/terminating-gateway-custom-and-tagged-addresses.envoy-1-18-x.golden (hunk 1 above)
Comment (freddygv) [reply agent/xds/testdata/listeners/terminating-gateway-custom-and-tagged-addresses.envoy-1-18-x.golden:35]: Doesn't look like it. Just tested this without it and got the same metrics names.

**Review by rboyer — APPROVED** [review]
LGTM

**Comment by hc-github-team-consul-core** [comment]
:cherries: If backport labels were added before merging, cherry-picking will start automatically.

To retroactively trigger a backport after merging, add backport labels and re-run https://circleci.com/gh/hashicorp/consul/387552.

**Comment by hc-github-team-consul-core** [comment]
:cherries::white_check_mark: Cherry pick of commit ae886136f115b1451b8e713f4446193ea44e6bde onto `release/1.10.x` succeeded!

**Pull request merged**